pom.xml

```diff
@@ -3273,7 +3273,7 @@
         <curator.version>2.7.1</curator.version>
         <commons-io.version>2.4</commons-io.version>
         <hadoop-client-api.artifact>hadoop-client</hadoop-client-api.artifact>
-        <hadoop-client-runtime.artifact>hadoop-client</hadoop-client-runtime.artifact>
+        <hadoop-client-runtime.artifact>hadoop-yarn-api</hadoop-client-runtime.artifact>
```

Comment:
Ahhh, this is a clever fix:

Instead of the `hadoop-2.7` profile resulting in a duplicate direct dependency on `hadoop-client`, we now just declare an explicit dependency on one of `hadoop-client`'s transitive dependencies (`hadoop-yarn-api` in this case). Anything which depends on `hadoop-client-runtime.artifact` must also depend on `hadoop-client-api.artifact`, so this doesn't end up changing the set of dependencies pulled in.

It looks like we didn't need to do that for `hadoop-client-minicluster.artifact` because that's only used in the `resource-managers/yarn` POM and that's already using Maven profiles to control the dependency selection (so the other workaround is less invasive in that context). In principle, though, I guess we could have changed that to some other transitive dep.

---

Could you maybe add a one or two line comment above these Hadoop 2.7 lines to explain what's going on? And maybe edit the comment at https://github.com/apache/spark/blob/d73562ed3635bb3454ac67029ca6541b30ae0c02/pom.xml#L251-L255 to reflect this change? This fix is clever but a little subtle, so I think a comment calling it out (and maybe mentioning SPARK-36835) might help future readers.

Reply:
Thanks for taking a look. Yes I think it's better to apply the same for `hadoop-client-minicluster.artifact. Let me try that, and perhaps we won't need the changes in YARN's pom.xml with this.

The side effect for this is seems to be that it affects the _distance_ of these dependencies to the root module and thus may make a difference when maven tries to resolve a dependency with multiple versions (see [here](https://maven.apache.org/guides/introduction/introduction-to-dependency-mechanism.html) for reference). I was using `hadoop-common` (which carries lots of dependencies) instead of `hadoop-yarn-api` and it was not able to compile.

Will update PR description and the comment in the above pom.xml.

Reply:
Actually it may not be so useful to change `hadoop-client-minicluster.artifact` since it is test scope while the other two are compile scope by default. For some reason it also changes `dev/deps/spark-deps-hadoop-2.7-hive-2.3` when I set it to something like `hadoop-mapreduce-client-jobclient`.